File: src/main/java/algo/twoPointers/LongestWordInDictionaryThroughDeleting_524.java
package algo.twoPointers;

import java.util.Arrays;
import java.util.List;

public class LongestWordInDictionaryThroughDeleting_524 {
  public static String findLongestWord(String s, List<String> dictionary) {
    String longest = "";
    var sorted = dictionary.stream().sorted().toList();
    for (String word : sorted) {
      int i = 0, j = 0;
      while (i < word.length() && j < s.length()) {
        if (word.charAt(i) == s.charAt(j)) {
          i++;
        }
        j++;
      }
      if (i == word.length()) {
        if (word.length() > longest.length()) {
          longest = word;
        }
      }
    }

    return longest;
  }

  public static void main(String[] args) {
    System.out.println(findLongestWord("abpcplea", List.of("ale","apple","monkey","plea")));
    System.out.println(findLongestWord("abpcplea", List.of("a", "b")));
    System.out.println(findLongestWord("abce", List.of("abe","abc")));
  }
}

File: src/main/kotlin/algo/twoPointers/LongestWordInDictionaryThroughDeleting_524.kt
package algo.twoPointers

class KtLongestWordInDictionaryThroughDeleting_524 {
    fun findLongestWord(s: String, dictionary: List<String>): String {
        var longestWord = ""
        val sorted = dictionary.sorted()
        sorted.forEach { word ->
            if (word.length > s.length) {
                return@forEach
            }
            var i = 0; var j = 0
            while (i < s.length && j < word.length) {
                if (s[i] == word[j]) {
                    j++
                }
                i++
            }
            // matched
            if (j == word.length) {
                if (word.length > longestWord.length) {
                    longestWord = word
                }
            }
        }
        return longestWord
    }
}